Added connection interface, database manager, generic singleton, tests, comments...

allanhaywood opened the pull request:
Added connection interface.
Added databasemanager.
Added generic singleton, along with some additional tests.
Made database manager a singleton.
Updated comments.

allanhaywood committed:
Added connection class interface.

allanhaywood committed:
Added databasemanager.

allanhaywood committed:
Merge remote-tracking branch 'origin/master' into master_DatabaseManager

allanhaywood committed:
Added generic singleton.
Added additional databasemanager constructor that takes a json file path.
Added test for the additional database manager constructor.
Updated comments.
Removed unecssary connection.cpp, renamed to iconnection.

joeyfreeman commented on PlotTalk/Src/jsonconnection.cpp:
```diff
@@ -35,7 +36,7 @@ QString JsonConnection::getPathToJson()
  * TODO: Change behavior for when a matching TvShow is not found, need to investigate best option.
  */
-TvShow JsonConnection::getTvShow(QString name)
+void JsonConnection::getTvShow(QString name, TvShow &tvShow)
```
Might be best to make the media item getters into bools that return false without modifying the passed in item if no match is found.

allanhaywood replied:
I was working on a bit of a re-design, which I just pushed, check out the changes. It exposes an easier to use interface, which isn't void anymore (this was before I read your comment).

What will probably be desired is throwing a known (ex. notfound) exception when a searched for tvshow is not found, since a return bool is not available to use anymore.

haleyminshell reviewed (approved):
I looked through the code and it looks good to me. Also ran through the unit tests and they passed.

allanhaywood committed:
Changed database manager to not implment iconnection, instead it exposes an easier to use interface
that will mirror the functionality of iconnection.
Added local store for TvShow objects in databasemanager. Required due to the above, and the need
for a long lived singleton to maintain the database objects (instead of the callers on the forms maintaining them).
Updated uml to match change.

allanhaywood committed:
Adding exception support and examples with a negative unit test.

allanhaywood committed:
Added comment to plot talk exceptions.

joeyfreeman reviewed (approved):
Successfully set up and ran the unit tests. Exceptions header looks good, I approve!
